File: 3948-maximum-number-of-subsequences-after-one-inserting/maximum-number-of-subsequences-after-one-inserting.cpp
class Solution {
public:
   long long numOfSubsequences(string s) {
    int n = s.length();

    vector<int> leftprec(n + 1, 0);   // prefix count of 'C'
    vector<int> rightprec(n + 1, 0);  // suffix count of 'C'

    vector<int> leftprel(n + 1, 0);   // prefix count of 'L'
    vector<int> rightprel(n + 1, 0);  // suffix count of 'L'

    vector<int> leftpret(n + 1, 0);   // prefix count of 'T'
    vector<int> rightpret(n + 1, 0);  // suffix count of 'T'

    // Build prefix counts
    for (int i = 0; i < n; i++) {
        leftprel[i + 1] = leftprel[i] + (s[i] == 'L' ? 1 : 0);
        leftprec[i + 1] = leftprec[i] + (s[i] == 'C' ? 1 : 0);
        leftpret[i + 1] = leftpret[i] + (s[i] == 'T' ? 1 : 0);
    }

    // Build suffix counts
    for (int i = n - 1; i >= 0; i--) {
        rightprel[i] = rightprel[i + 1] + (s[i] == 'L' ? 1 : 0);
        rightprec[i] = rightprec[i + 1] + (s[i] == 'C' ? 1 : 0);
        rightpret[i] = rightpret[i + 1] + (s[i] == 'T' ? 1 : 0);
    }

    long long ans = 0;
    long long maxxin = 0;   // For inserting 'C'
    long long maxxint = 0;  // For inserting 'T'
    long long maxxinl = 0;  // For inserting 'L'

    for (int i = 0; i < n; i++) {
        if (s[i] == 'C') {
            int left = leftprel[i];       // 'L's before this 'C'
            int right = rightpret[i];     // 'T's after this 'C'
            ans += 1LL * left * right;

            // Insert 'L' before this 'C'  contributes right 'T's
            maxxinl += right;

            // Insert 'T' after this 'C'  contributes left 'L's
            maxxint += left;
        }

        // Insert 'C' at position i: left 'L's × right 'T's
        int leftl = leftprel[i];
        int rightt = rightpret[i];
        maxxin = max(maxxin, 1LL * leftl * rightt);
    }


    return ans + max({maxxin, maxxinl, maxxint});
}

};

File: 3948-maximum-number-of-subsequences-after-one-inserting/maximum-number-of-subsequences-after-one-inserting.js
/**
 * @param {string} s
 * @return {number}
 */
var numOfSubsequences = function(s) {
    let n=s.length;
       let leftprec = new Array(n+1).fill(0);   // count of 'L' up to index i
    let rightprec= new Array(n+1).fill(0);
       let leftprel = new Array(n+1).fill(0);   // count of 'L' up to index i
    let rightprel= new Array(n+1).fill(0);
       let leftpret = new Array(n+1).fill(0);   // count of 'L' up to index i
    let rightpret= new Array(n+1).fill(0);
    
        for(let i=0;i<s.length;i++)
        {
                  if(s[i]=='L')
                  {
                      leftprel[i+1]=leftprel[i]+1; 
                      
                    //   console.log("k");
                  }else
                   leftprel[i+1]=leftprel[i]; 

                       if(s[i]=='C')
                  {
                      leftprec[i+1]=leftprec[i]+1; 
                      
                    //   console.log("k");
                  }else
                   leftprec[i+1]=leftprec[i]; 

                //        if(s[i]=='t')
                //   {
                //       leftprel[i+1]=leftprel[i]+1; 
                      
                //     //   console.log("k");
                //   }else
                //    leftprel[i+1]=leftprel[i]; 
              
                

        }
        for(let i=s.length-1;i>=0;i--)
        {
                  if(s[i]=='T')
                  {
                      rightpret[i]=rightpret[i+1]+1; 
                  }else
                   rightpret[i]=rightpret[i+1]; 

                    if(s[i]=='C')
                  {
                      rightprec[i]=rightprec[i+1]+1; 
                  }else
                   rightprec[i]=rightprec[i+1];

                    if(s[i]=='L')
                  {
                      rightprel[i]=rightprel[i+1]+1; 
                  }else
                   rightprel[i]=rightprel[i+1]; 

        }
        var ans=0;
        var maxxin=0;
        var maxxint=0;
        var maxxinl=0;
       for(let i=0;i<s.length;i++)
        {
                 if(s[i]==='C')
                 {
                    
                     
                         let left=leftprel[i];
                         let right=rightpret[i];
                        
                        //insert l
                         maxxinl+=(right);
                        //  insert t
                        maxxint+=(left);
                         ans+=(left*right);
                        //  console.log(ans);
                 }
                 //insert c
                   let leftl=leftprel[i];
                         let rightt=rightpret[i];
                         maxxin=Math.max(maxxin,leftl*rightt);
                //  //insert l
                //  let temc=leftprec[i];
                //  let temt=rightpret[i];
                //   maxxin=Math.max(maxxin,temc*temt);
                //   //insert t
                //    let teml=leftprel[i];
                //  let temcl=rightprec[i];
                //   maxxin=Math.max(maxxin,temcl*teml);



        }
        // var maxxin=0;
        //    for(let i=0;i<s.length;i++)
        // {
        //           //insert L
        //            let left=leftpre[i];
        //                  let right=rightpre[i];
        //            let tem=left*right;
        //            maxxin=Math.max(maxxin,tem);
        // }

        // console.log(ans);
        console.log(maxxin+" "+maxxinl+" "+maxxint);
        return ans+Math.max(maxxin,maxxinl,maxxint);
};